### PR TITLE
Add ContainerType and default resource quotas

### PR DESCRIPTION
**Pull request by Techassi**
Implements #596

### CODE REVIEW

**Comment by Techassi** [comment]
I incorporated the code from @sbernauer solution. Additionally the presence of CPU and memory limits is validated when building the Pod spec.

**Comment by Techassi** [comment]
This PR also adds `ContainerType` which might be somewhat out of scope, but this topic should be discussed as it might prove useful in the future. Having container types attached to each container would allow us to more easily specify custom code based on the type of container, like resource quota limits or something else entirely.

Depending on the discussion, the affected code could be removed from this PR and moved to a separate PR.

File: src/container_type.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+/// Describes the type of container. The types are:
+///
+/// - `Main`: This type of container runs the main appliction, like Superset or
+///    NiFi
+/// - `Init`: This type of container runs only once at the start of the product
+///    cluster and executes initial tasks like database migrations or insertion
+///    of data
+/// - `Sidecar`: Sidecar containers run alongside the main containers for
+///    additional tasks
+#[derive(Clone, Debug, Default)]
```
Comment (sbernauer) [inline src/container_type.rs:14]: If we go with the ContainerType I would make it more explicit, e.g. have
* Product (main)
* copy files (init)
* text replacement (init)
* create key and truststore (init)
* logging (sidecar)
* metrics (sidecar)
* healthcheck (sidecar)
* ...

File: src/container_type.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    Sidecar,
+}
+
+impl From<ContainerType> for ResourceRequirements {
+    fn from(container_type: ContainerType) -> Self {
+        Self::from(&container_type)
+    }
+}
+
+impl From<&ContainerType> for ResourceRequirements {
```
Comment (sbernauer) [inline src/container_type.rs:28]: We really only want to set default resources on init containers

**Review by sbernauer — CHANGES_REQUESTED** [review]
In general I would prefer to not include the ContainerType bits in this PR

File: src/builder/pod/mod.rs [hunk 3]
```diff
@@ -367,8 +394,57 @@ impl PodBuilder {
...
+        Ok(PodTemplateSpec {
+            metadata: self.metadata.clone(),
+            spec: Some(self.build_spec()?),
+        })
+    }
+
+    /// Returns if any of the containers is missing the resource quota limit
+    /// for `key`. If so, [`Some(container_name)`] is returned, otherwise
+    /// [`None`].
+    fn is_missing_limit_for(&self, key: &str) -> Option<String> {
```
Comment (sbernauer) [inline src/builder/pod/mod.rs:425]:
```suggestion
    fn check_resource_limit_set(&self, key: &str) -> OperatorResult<()> {
```

File: src/builder/pod/mod.rs [hunk 4]
```diff
@@ -367,8 +394,57 @@ impl PodBuilder {
...
+            tracing::warn!("Tried building a PodTemplate for a PodBuilder with a status, the status will be ignored...");
+        }
+
+        Ok(PodTemplateSpec {
+            metadata: self.metadata.clone(),
+            spec: Some(self.build_spec()?),
+        })
+    }
+
+    /// Returns if any of the containers is missing the resource quota limit
```
Comment (sbernauer) [inline src/builder/pod/mod.rs:422]:
```suggestion
    /// Returns if any of the containers is missing the given resource limit
```

File: src/builder/pod/container.rs [hunk 5]
```diff
@@ -223,14 +227,25 @@ impl ContainerBuilder {
         self
     }
 
+    /// Sets the container type.
+    pub fn container_type(&mut self, container_type: ContainerType) -> &mut Self {
```
[the remark this answers is not in the document]
Comment (Techassi) [reply src/builder/pod/container.rs]: Agreed. This will be a mandatory parameter in the `ContainerBuilder::new()` function.

File: src/builder/pod/mod.rs (hunk 3 above)
Comment (Techassi) [reply src/builder/pod/mod.rs]: Changed in 5621b6f

File: src/builder/pod/mod.rs (hunk 4 above)
Comment (Techassi) [reply src/builder/pod/mod.rs]: Changed in 5621b6f

File: src/container_type.rs (hunk 1 above)
Comment (Techassi) [reply src/container_type.rs]: The exact variants should be discussed in the upcoming separate PR. And yes we can be more explicit about the operations the container executes.

File: src/container_type.rs (hunk 2 above)
Comment (Techassi) [reply src/container_type.rs]: Yeah makes sense!

File: rust-toolchain.toml [hunk 6]
```diff
@@ -0,0 +1,2 @@
+[toolchain]
```
Comment (sbernauer) [inline rust-toolchain.toml:1]: LGTM. CC @razvan because you know the templating stuff pretty good. Maybe we want to template this here.

File: src/builder/pod/mod.rs [hunk 7]
```diff
@@ -161,7 +162,33 @@ impl PodBuilder {
...
 
-    pub fn add_init_container(&mut self, container: Container) -> &mut Self {
+    /// Add the given init container.
+    /// If no resources are set, we set a default limit of 10m CPU and 128Mi
+    /// memory. Request values are set to the same values as the limits.
+    pub fn add_init_container(&mut self, mut container: Container) -> &mut Self {
+        // https://github.com/stackabletech/issues/issues/368:
+        // We only set default limits on *init* containers, as they normally
+        // simply copy stuff around, do some text replacement or, at a maximum,
+        //  create a tls truststore.These operations should normally complete
```
Comment (sbernauer) [inline src/builder/pod/mod.rs:172]:
```suggestion
        // create a tls truststore.These operations should normally complete
```

File: src/builder/pod/mod.rs [hunk 8]
```diff
@@ -281,12 +327,27 @@ impl PodBuilder {
...
+    ///     apimachinery::pkg::api::resource::Quantity
+    /// };
+    ///
     /// let pod = PodBuilder::new()
     ///     .metadata_default()
     ///     .add_container(
     ///         ContainerBuilder::new("container")
     ///             .unwrap()
     ///             .add_volume_mount("listener", "/path/to/volume")
+    ///             .resources(ResourceRequirements {
```
Comment (sbernauer) [inline src/builder/pod/mod.rs:341]: Nit: Maybe we can have a fn on ContainerBuilder similar to `fn set_resources(cpu_limit: Quantity, memory_limit: Quantity, cpu_request:: Option<Quantity>, memory_request:: Option<Quantity>)`

File: src/commons/resources.rs [hunk 9]
```diff
@@ -318,6 +319,107 @@ impl<T, K> Into<ResourceRequirements> for Resources<T, K> {
     }
 }
 
+#[derive(Debug, thiserror::Error)]
+#[error("missing {resource_key} resource {resource_policy} for container {container_name}")]
+pub struct ResourceRequirementsPolicyError {
```
Comment (sbernauer) [inline src/commons/resources.rs:324]: I struggle a bit with the word `Policy`, as it does not seem to be a k8s term (correct me if I'm wrong).
Finding an alternative is pretty hard. WDYT of simply calling it `ResourceRequirements` instead of `ResourceRequirementsPolicy` (through the PR)?

File: src/commons/resources.rs [hunk 10]
```diff
@@ -318,6 +319,107 @@ impl<T, K> Into<ResourceRequirements> for Resources<T, K> {
...
+
+                if !requests.contains_key(resource) {
+                    return Err(ResourceRequirementsPolicyError {
+                        container_name: self.name.clone(),
+                        resource_key: resource.into(),
+                        resource_policy: policy,
+                    });
+                }
+            }
+            ResourceRequirementsPolicy::Claims => todo!(),
```
Comment (sbernauer) [inline src/commons/resources.rs:404]: Marker to not forget.

File: src/commons/resources.rs [hunk 11]
```diff
@@ -318,6 +319,107 @@ impl<T, K> Into<ResourceRequirements> for Resources<T, K> {
...
+                        resource_key: resource.into(),
+                        resource_policy: policy,
+                    });
+                }
+            }
+            ResourceRequirementsPolicy::Requests => {
+                let requests = rr
+                    .requests
+                    .as_ref()
+                    .ok_or(ResourceRequirementsPolicyError {
```
Comment (sbernauer) [inline src/commons/resources.rs:390]: :+1:

**Review by sbernauer — CHANGES_REQUESTED** [review]
Overall direction seems good.
I had the idea, that it would be good to (also?) check the resources in the cluster resources code. 
I think we should extend the `ClusterResource` trait https://github.com/stackabletech/operator-rs/blob/5478d90da51ee099360f2ccd003119f62b56c806/src/cluster_resources.rs#L46 with a fn to get the *optional* PodSpec and implement it for Job, DaemonSet and Statefulset.
Something like 
```
    fn pod_spec(self) -> Option<PodSpec> {
        None
    }
```
Afterwards than make assertions on the Pod here https://github.com/stackabletech/operator-rs/blob/5478d90da51ee099360f2ccd003119f62b56c806/src/cluster_resources.rs#L406.

File: src/commons/resources.rs (hunk 9 above)
Comment (Techassi) [reply src/commons/resources.rs:324]: The thing is that the term `ResourceRequirements` already exists (Imported from `k8s_openapi` and used across `operator-rs`. The term describes the complete struct, which contains the fields `requests`, `limits` and `claims`. One alternative name could be `ResourceRequirementsType`.

File: src/commons/resources.rs (hunk 10 above)
Comment (Techassi) [reply src/commons/resources.rs:404]: How do we handle claims? Do we need to do any validation here?

File: rust-toolchain.toml (hunk 6 above)
Comment (razvan) [reply rust-toolchain.toml:1]: This is already in operator-templating.

**Review by razvan — COMMENTED** [review]
Sorry for not commenting sooner. Now that I have some experience with failed tests due to resource constraints I find that failing to create pods /statefulsets/etc. when resource requests are missing is confusing:

* The error messages get lost in the reconciliation log stream
* As a user/admin I watch K8S objects and I am surprised when they are not there.
* I have to watch logs for these errors explicitly and there is no good/easy  tooling for that now.
* In case of such failures, K8S persists the message in the status of these objects. I can use kubectl to inspect them.

TL;DR: I propose to not fail when resource requests are missing. but maybe just warn.

**Comment by razvan** [comment]
One other reason not to fail in this case is because it feels like such a special case and inconsistent.

There so many other reasons why creating K8S objects (pods/statefulsets) is not "semantically correct" yet we still create them.

**Comment by razvan** [comment]
And one last reason why I don't like this approach: the status fields of the clusters is not updated to explain why the cluster is in an "incomplete" state.

File: src/commons/resources.rs (hunk 10 above)
Comment (sbernauer) [reply src/commons/resources.rs:404]: Nope, I'm not aware of any claim usages. I would suggest to not support it for now

File: rust-toolchain.toml (hunk 6 above)
Comment (sbernauer) [reply rust-toolchain.toml:1]: Yeah, but looking at https://github.com/stackabletech/operator-templating/blob/main/repositories.yaml only for the product operator, is that correct? We are in operator-rs here.
And: If the answer is "Not worth the effort" that's totally fine by me.

File: src/commons/resources.rs (hunk 10 above)
Comment (Techassi) [reply src/commons/resources.rs:404]: So should we remove it from the enum completely or should we keep the `todo!()` macro (or print out an error)

File: src/builder/pod/mod.rs (hunk 8 above)
Comment (Techassi) [reply src/builder/pod/mod.rs:341]: Agreed, will implement this.

File: src/commons/resources.rs (hunk 10 above)
Comment (sbernauer) [reply src/commons/resources.rs:404]: i would remove the enum variant

File: src/builder/pod/mod.rs (hunk 8 above)
Comment (Techassi) [reply src/builder/pod/mod.rs:342]: Added in d48d06c

File: src/commons/resources.rs (hunk 9 above)
Comment (Techassi) [reply src/commons/resources.rs]: Renamed in d48d06c

File: src/commons/resources.rs (hunk 10 above)
Comment (Techassi) [reply src/commons/resources.rs]: Removed in d48d06c

**Comment by Techassi** [comment]
Since d48d06c, the `PodBuilder` will only log a warning when building the Pod / PodSpec and resource limits.

**Comment by Techassi** [comment]
> Overall direction seems good. I had the idea, that it would be good to (also?) check the resources in the cluster resources code. I think we should extend the `ClusterResource` trait
> 
> https://github.com/stackabletech/operator-rs/blob/5478d90da51ee099360f2ccd003119f62b56c806/src/cluster_resources.rs#L46
> with a fn to get the _optional_ PodSpec and implement it for Job, DaemonSet and Statefulset.
> Something like
> 
> ```
>     fn pod_spec(self) -> Option<PodSpec> {
>         None
>     }
> ```
> 
> Afterwards than make assertions on the Pod here
> 
> https://github.com/stackabletech/operator-rs/blob/5478d90da51ee099360f2ccd003119f62b56c806/src/cluster_resources.rs#L406
> .

Do we really need another check here? All Pods/PodSpecs built by the `PodBuilder` should contain resource requirements, and if not, a warning will be logged.

File: rust-toolchain.toml (hunk 6 above)
Comment (Techassi) [reply rust-toolchain.toml:1]: Okay how should we proceed then? I think manually updating this should be easy, but it should probably be documented (or be part of the post/pre-release tasks).

File: src/builder/pod/container.rs [hunk 12]
```diff
@@ -223,6 +223,104 @@ impl ContainerBuilder {
         self
     }
 
+    /// Sets the CPU limit and optionally request quantity for this container.
+    /// This overwrites previously defined quantitities for both limit and
+    /// request. Internally this uses [`self.with_resource`].
+    pub fn with_cpu(&mut self, limit: Quantity, request: Option<Quantity>) -> &mut Self {
```
[the remark this answers is not in the document]
Comment (Techassi) [reply src/builder/pod/container.rs:229]: Actually, I'm experimenting with such a builder right now. Regarding the misuse from above: Separating the CPU limit from request (with two builder methods) would make this more clear in my opinion.

**Comment by sbernauer** [comment]
> Do we really need another check here? All Pods/PodSpecs built by the PodBuilder should contain resource requirements, and if not, a warning will be logged.

Nope, current state is fine for me :+1: It's only about warnings, nothing more

File: src/builder/pod/container.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply src/builder/pod/container.rs]: Implemented in b81eeb6

**Comment by Techassi** [comment]
I implemented the `ResourceRequirementsBuilder` which should support all the stuff we need right now. Next step would to add some doc comments. The PR should then be ready to merge.

File: src/builder/pod/mod.rs [hunk 13]
```diff
@@ -367,7 +437,62 @@ impl PodBuilder {
...
+    }
+
     fn build_spec(&self) -> PodSpec {
+        // We don't hard error here, because if we do, the StatefulSet (for
+        // example) doesn't show up at all. Instead users then need to comb
+        // through the logs to find the error. That's why we opted to just
+        // throw a warning which will get displayed in the Kubernetes
+        // status.
+
+        if let Err(err) = self.check_container_resource(ResourceRequirementsType::Limits, "cpu") {
```
Comment (sbernauer) [inline src/builder/pod/mod.rs:487]: I would rather build the PodSpec, and than use
```
impl ResourceRequirementsExt for PodSpec {
    fn check_resource_requirement(
        &self,
        rr_type: ResourceRequirementsType,
        resource: &str,
    ) -> Result<(), ResourceRequirementsError> {
        for container in &self.containers {
            container.check_resource_requirement(rr_type, resource)?
        }

        Ok(())
    }
}
```

We also need to check this in `ClusterResource::add`, otherwise it was useless to extend the Trait ;)
```
        if let Some(pod_spec) = resource.pod_spec() {
            if let Err(err) =
                pod_spec.check_resource_requirement(ResourceRequirementsType::Limits, "cpu")
            {
                warn!("{}", err)
            }
            if let Err(err) =
                pod_spec.check_resource_requirement(ResourceRequirementsType::Limits, "memory")
            {
                warn!("{}", err)
            }
        }
```

File: src/builder/pod/mod.rs [hunk 14]
```diff
@@ -367,7 +437,62 @@ impl PodBuilder {
         self
     }
 
+    /// Consumes the Builder and returns a constructed [`Pod`]
+    pub fn build(&self) -> OperatorResult<Pod> {
```
[the remark this answers is not in the document]
Comment (Techassi) [reply src/builder/pod/mod.rs:441]: Yes I moved the private method `build_spec` to the end of the `impl` block. I like grouping public and private methods and usually I put private methods at the end.

File: src/commons/resources.rs [hunk 15]
```diff
@@ -318,6 +319,123 @@ impl<T, K> Into<ResourceRequirements> for Resources<T, K> {
...
+        &self,
+        rr_type: ResourceRequirementsType,
+        resource: &str,
+    ) -> Result<(), ResourceRequirementsError>;
+
+    /// Returns wether the implementor has a [`ResourceRequirementsType`] set
+    /// for a `resource`.
+    fn has_resource_requirement(&self, rr_type: ResourceRequirementsType, resource: &str) -> bool {
+        self.check_resource_requirement(rr_type, resource)
+            .map_or(false, |_| true)
```
[the remark this answers is not in the document]
Comment (Techassi) [reply src/commons/resources.rs]: Changed in 18c98da

File: src/builder/resources.rs [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+impl ResourceRequirementsBuilder<state::Initial> {
+    pub fn new() -> Self {
+        ResourceRequirementsBuilder::default()
+    }
+
+    pub fn with_cpu_limit(
+        self,
+        limit: Quantity,
```
[the remark this answers is not in the document]
Comment (Techassi) [reply src/builder/resources.rs]: Yes this is nice, changed in 18c98da

File: src/builder/resources.rs [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+use k8s_openapi::{
+    api::core::v1::ResourceRequirements, apimachinery::pkg::api::resource::Quantity,
+};
+
+mod state {
+    #[derive(Debug, Default)]
+    pub struct Initial {}
+    pub struct MissingCpuLimit {}
+    pub struct MissingMemLimit {}
+    pub struct Final {}
```
[the remark this answers is not in the document]
Comment (Techassi) [reply src/builder/resources.rs]: Changed in 18c98da

File: src/builder/resources.rs [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+            Some(limits)
+        } else {
+            None
+        };
+
+        let requests = if !requests.is_empty() {
+            Some(requests)
+        } else {
+            None
+        };
```
[the remark this answers is not in the document]
Comment (Techassi) [reply src/builder/resources.rs]: This pattern is also used [here](https://github.com/stackabletech/operator-rs/blob/main/src/commons/resources.rs#L302-L317). I just adjusted the ordering, otherwise leaving as is.

File: src/builder/pod/container.rs [hunk 19]
```diff
@@ -223,6 +218,11 @@ impl ContainerBuilder {
         self
     }
 
+    pub fn with_resources(&mut self, resources: ResourceRequirements) -> &mut Self {
```
[the remark this answers is not in the document]
Comment (Techassi) [reply src/builder/pod/container.rs:221]: I feel like builders should prefix all methods with either `add` or `with`. This makes it super obvious that the user is dealing with a builder.

File: src/builder/pod/mod.rs (hunk 13 above)
Comment (Techassi) [reply src/builder/pod/mod.rs:473]: Added in 597e751

File: src/cluster_resources.rs [hunk 20]
```diff
@@ -58,6 +61,10 @@ pub trait ClusterResource:
     fn maybe_mutate(self, _strategy: &ClusterResourceApplyStrategy) -> Self {
         self
     }
+
+    fn pod_spec(self) -> Option<PodSpec> {
```
Comment (sbernauer) [inline src/cluster_resources.rs:65]: Maybe this helps avoiding unnecessary clones introduced in 597e751e37540dd7d4c3192b7d786043048f0f9c
```suggestion
    fn pod_spec(&self) -> Option<PodSpec> {
```

Comment (Techassi) [reply src/cluster_resources.rs]: 3fc8f1d takes `&self` and clones inside the method.

File: src/builder/resources.rs [hunk 21]
```diff
@@ -0,0 +1,218 @@
+use std::{collections::BTreeMap, marker::PhantomData};
```
[the remark this answers is not in the document]
Comment (Techassi) [reply src/builder/resources.rs]: Moved in 3fc8f1d

File: src/builder/pod/mod.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (sbernauer) [reply src/builder/pod/mod.rs:442]: Kind of makes sense ;) On the other hand having build() at the bottom aligns better with the code flow. LGTM ;)

File: src/cluster_resources.rs [hunk 22]
```diff
@@ -58,6 +61,10 @@ pub trait ClusterResource:
     fn maybe_mutate(self, _strategy: &ClusterResourceApplyStrategy) -> Self {
         self
     }
+
+    fn pod_spec(&self) -> Option<PodSpec> {
```
Comment (sbernauer) [inline src/cluster_resources.rs:65]:
```suggestion
    fn pod_spec(&self) -> Option<&PodSpec> {
```

File: src/cluster_resources.rs [hunk 23]
```diff
@@ -155,7 +162,13 @@ impl ClusterResource for Service {}
 impl ClusterResource for ServiceAccount {}
 impl ClusterResource for RoleBinding {}
 impl ClusterResource for Secret {}
-impl ClusterResource for Job {}
+
+impl ClusterResource for Job {
+    fn pod_spec(&self) -> Option<PodSpec> {
+        self.spec.clone().unwrap_or_default().template.spec
+    }
```
Comment (sbernauer) [inline src/cluster_resources.rs:169]:
```suggestion
    fn pod_spec(&self) -> Option<&PodSpec> {
        self.spec
            .as_ref()
            .and_then(|spec| spec.template.spec.as_ref())
    }
```

File: src/cluster_resources.rs [hunk 24]
```diff
@@ -171,7 +184,12 @@ impl ClusterResource for StatefulSet {
             | ClusterResourceApplyStrategy::NoApply => self,
         }
     }
+
+    fn pod_spec(&self) -> Option<PodSpec> {
+        self.spec.clone().unwrap_or_default().template.spec
+    }
```
Comment (sbernauer) [inline src/cluster_resources.rs:190]:
```suggestion
    fn pod_spec(&self) -> Option<&PodSpec> {
        self.spec
            .as_ref()
            .and_then(|spec| spec.template.spec.as_ref())
    }
```

File: src/cluster_resources.rs [hunk 25]
```diff
@@ -206,6 +224,10 @@ impl ClusterResource for DaemonSet {
...
```
Comment (sbernauer) [inline src/cluster_resources.rs:230]:
```suggestion
    fn pod_spec(&self) -> Option<&PodSpec> {
        self.spec
            .as_ref()
            .and_then(|spec| spec.template.spec.as_ref())
    }
```

**Review by sbernauer — COMMENTED** [review]
This way we don't need to clone the PodSpec (which is quite big) for *every* reconciliation loop

File: src/cluster_resources.rs (hunk 20 above)
Comment (sbernauer) [reply src/cluster_resources.rs]: Continued below with suggestion

File: src/commons/resources.rs [hunk 26]
```diff
@@ -318,6 +319,135 @@ impl<T, K> Into<ResourceRequirements> for Resources<T, K> {
...
 
+#[derive(Debug, thiserror::Error)]
+#[error("missing {resource_key} resource {resource_type} for container {container_name}")]
+pub struct ResourceRequirementsError {
+    resource_type: ResourceRequirementsType,
+    container_name: String,
+    resource_key: String,
+}
+
+/// [`ResourceRequirementsType`] dsscribes the available resource requirement
```
Comment (sbernauer) [inline src/commons/resources.rs:330]:
```suggestion
/// [`ResourceRequirementsType`] describes the available resource requirement
```

File: src/builder/pod/resources.rs [hunk 27]
```diff
@@ -0,0 +1,222 @@
...
+                "setting resource '{}' directly is discouraged - use provided methods instead",
+                resource
+            );
+            return self;
+        }
+
+        let resource = resource.to_string();
+
+        if self.other.contains_key(&resource) {
+            warn!("resource '{}' already set, not overwriting", resource);
```
[the remark this answers is not in the document]
Comment (Techassi) [reply src/builder/pod/resources.rs]: Changed to FIXME in 73f0fe0. We should tackle this after our container purpose discussion and maybe in the PR where we introduce the Type State Pattern in the `ContainerBuilder`.

File: src/commons/resources.rs (hunk 26 above)
Comment (Techassi) [reply src/commons/resources.rs]: Fixed in 73f0fe0

File: src/builder/pod/container.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply src/builder/pod/container.rs]: Reverted in 59049cc

File: src/builder/pod/mod.rs [hunk 28]
```diff
@@ -161,7 +164,43 @@ impl PodBuilder {
...
+        // create a tls truststore.These operations should normally complete
+        // in <= 1s, so worst-case the Pod will take 1-2s longer to start up
+        // when the default is too low. However, things are different with
+        // sidecars, where e.g. a bundle builder, metric collector or a vector
+        // log sidecar can be overloaded and slow down operations or cause
+        // missing data, e.g. metrics or logs. Hence we don't apply any defaults
+        // for sidecars, product operators have to explicitly make a decision
+        // on what the resource limits should be.
+
+        // TODO (Techassi): These defaults should not live here and should
```
[the remark this answers is not in the document]
Comment (Techassi) [reply src/builder/pod/mod.rs]: Changed to FIXME in 73f0fe0. We should tackle this after our container purpose discussion and maybe in the PR where we introduce the Type State Pattern in the `ContainerBuilder`.

**Comment by Techassi** [comment]
> Only thing remaining is the `with_resources` discussion and adding a warning when mem request < limit

Shouldn't we add a warning when request > limit. And also for both CPU and memory?

**Comment by sbernauer** [comment]
Only for memory, not for CPU. CPU can be easily throttled when the node does not have enough cpu resources. When the pod uses more memory than it requested it needs to be killed to get memory back.

File: src/builder/pod/mod.rs (hunk 28 above)
[the remark this answers is not in the document]
Comment (sbernauer) [reply src/builder/pod/mod.rs]: As we skip the container purpose for now, I don't see a better place to put this (where we now we deal with init containers only) ;)
Fine for me to leave it in anyway

File: src/builder/pod/mod.rs [hunk 29]
```diff
@@ -390,31 +499,40 @@ impl PodBuilder {
...
             ..PodSpec::default()
+        };
+
+        // We don't hard error here, because if we do, the StatefulSet (for
+        // example) doesn't show up at all. Instead users then need to comb
+        // through the logs to find the error. That's why we opted to just
+        // throw a warning which will get displayed in the Kubernetes
+        // status. Additionally the Statefulset will have events describing the
+        // actual problem.
+
```
Comment (sbernauer) [inline src/builder/pod/mod.rs:510]: I would prefer something like this, as this is more compact (same in cluster_resources)
```
        pod_spec
            .check_resource_requirement(ResourceRequirementsType::Limits, "cpu")
            .unwrap_or_else(|err| warn!("{err}"));
        pod_spec
            .check_resource_requirement(ResourceRequirementsType::Limits, "memory")
            .unwrap_or_else(|err| warn!("{err}"));
        pod_spec
            .check_limit_to_request_ratio(&ComputeResource::Cpu, LIMIT_REQUEST_RATIO_CPU)
            .unwrap_or_else(|err| warn!("{err}"));
        pod_spec
            .check_limit_to_request_ratio(&ComputeResource::Memory, LIMIT_REQUEST_RATIO_MEMORY)
            .unwrap_or_else(|err| warn!("{err}"));
```

File: src/commons/resources.rs [hunk 30]
```diff
@@ -318,17 +326,210 @@ impl<T, K> Into<ResourceRequirements> for Resources<T, K> {
...
+}
+
+/// [`ResourceRequirementsType`] describes the available resource requirement
+/// types. The user can set limits, requests and claims. This enum makes it
+/// possible to check if containers set one or more of these types.
+#[derive(Copy, Clone, Debug, Display, PartialEq, Eq, PartialOrd, Ord)]
+#[strum(serialize_all = "lowercase")]
+pub enum ResourceRequirementsType {
+    Limits,
+    Requests,
```
Comment (sbernauer) [inline src/commons/resources.rs:357]:
```suggestion
    Requests,
    Limits,
```

**Review by sbernauer — CHANGES_REQUESTED** [review]
LGTM % final comments

File: src/commons/resources.rs (hunk 30 above)
Comment (Techassi) [reply src/commons/resources.rs]: Changed in f5a7a7c

File: src/builder/pod/mod.rs (hunk 29 above)
Comment (Techassi) [reply src/builder/pod/mod.rs:510]: Changed in f5a7a7c

Comment (sbernauer) [reply src/builder/pod/mod.rs:510]: Please also for cluster_resources code :)

**Review by sbernauer — APPROVED** [review]
LGTM % last comment (also change cluster_resources)
Thanks for driving this!

File: src/builder/pod/mod.rs (hunk 29 above)
Comment (Techassi) [reply src/builder/pod/mod.rs:510]: Updated `cluster_resources` in e57a117
